InstructionDeleter rewrite and migrate passes

atrick opened the pull request:
This fixes correctness bugs that are blocking other optimization bug fixes...

Clarify the API. Make it suitable for use everywhere in the compiler. We should try to standardize on it and allow it to do the OSSA fixup more often.

Add InstructionDeleter::updatingIterator() factory so we never normally need to use InstModCallbacks.

Fix bugs in which notifyWillBeDeleted() was being called on invalid SIL. The bugs are easily exposed just by removing copy_value side effects, but that will be in the follow-up commit.

Call notifyWillBeDeleted() only when identifying new dead instructions that the client may not know about. Give the client control over force-deleting instructions. When doing its own lifetime fixups, the client may force-delete a set of related instructions. Invoking callbacks for these force-deleted instructions is wrong.

TODO: partial_apply support is only partial. I disabled the buggy
cases. This should be easy to fix but requires designing some
InstructionDeleter test cases.

atrick committed:
Add Operand::isConsuming().

Use this whenever we're assuming an owned value.

atrick committed:
InstructionDeleter rewrite

Clarify the API. Make it suitable for use everywhere in the
compiler. We should try to standardize on it and allow it to do the
OSSA fixup more often.

Add InstructionDeleter::updatingIterator() factory so we never
normally need to use InstModCallbacks.

Fix bugs in which notifyWillBeDeleted() was being called on invalid
SIL. The bugs are easily exposed just by removing copy_value side
effects, but that will be in the follow-up commit.

Call notifyWillBeDeleted() only when identifying new dead instructions
that the client may not know about. Give the client control over
force-deleting instructions. When doing its own lifetime fixups, the
client may force-delete a set of related instructions. Invoking
callbacks for these force-deleted instructions is wrong.

TODO: partial_apply support is only partial. I disabled the buggy
cases. This should be easy to fix but requires designing some
InstructionDeleter test cases.

atrick committed:
Rewrite instruction deletion logic in many passes

Fix innumerable latent bugs with iterator invalidation and callback invocation.

Removes dead code earlier and chips away at all the redundant copies the compiler generates.

atrick commented:
@swift-ci test

gottesmm commented on include/swift/SIL/SILValue.h:
```diff
@@ -1037,6 +1037,8 @@ class Operand {
   /// guaranteed scope.
   bool isLifetimeEnding() const;
 
+  bool isConsuming() const;
```
Quick drive by (need to read more). But this needs a comment to express the difference with isLifetimeEnding().

eeckstein reviewed (commented):
Basically looks great!
I especially like that InstructionDeleter provides a solution for the iterator invalidation problem (though I have some suggestions for that) to reduce the use of InstModCallbacks.

I have a few comments about comments (meta-comments?), which all summarize to: There are a lot of comments describing all the side effects, details, etc. This is good (though it's a bit scary that so much documentation is needed to correctly use the API).
But what's missing are a few sentences what the InstructionDeleter and its methods actually do. IMO a perfect API just needs at most a few sentences to be able to use it correctly.
For me the InstructionDeleter is a leaky abstraction, because I cannot use it without looking into its implementation.

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -59,39 +59,53 @@ class InstModCallbacks {
-  /// caller can via the callback gather state about the instruction that will
-  /// be deleted. As an example, see InstructionDeleter::deleteInstruction() in
-  /// InstOptUtils.cpp.
+  /// TODO: When zombie instructions are implemented at the module level, we
```
Updated the comment. But rewriting all the clients is another task.
```
  /// TODO: Now that instructions deletion can be delayed via
  /// SILModule::scheduleForDeletion(); there's no longer a good use case for
  /// calling eraseFromParent() within this callback. Rewrite all clients
  /// without doing the instruction deletion within the callback.
```

atrick commented:
> I especially like that InstructionDeleter provides a solution for the iterator invalidation problem (though I have some suggestions for that) to reduce the use of InstModCallbacks.
> 
> I have a few comments about comments (meta-comments?), which all summarize to: There are a lot of comments describing all the side effects, details, etc. This is good (though it's a bit scary that so much documentation is needed to correctly use the API).
> But what's missing are a few sentences what the InstructionDeleter and its methods actually do. IMO a perfect API just needs at most a few sentences to be able to use it correctly.
> For me the InstructionDeleter is a leaky abstraction, because I cannot use it without looking into its implementation.

This is all true. This PR doesn't provide any new abstraction. It does has a safe/dumb `deleteIfDead` API that you can always reach for. That's still much more powerful than `recursivelyDeleteTriviallyDeadInstructions` because it works in OSSA code and on scoped instructions.

There are lots of variations on how passes want to go about deleting. Sometimes SIL is in an invalid state while instructions are being deleted. Sometimes a set of dependent instructions need to be deleted, but they individually have side effects. Sometimes new instructions need to be inserted for OSSA cleanup. First all that complexity needs to be surfaced in one central location before each of the use cases can be rewritten. I know this because I spent weeks debugging the implicit assumptions being made in code that manually deletes instructions.

The primary goal of this PR is to fix bugs that are introduced by adding callbacks to the InstructionDeleter. In the course of fixing that we also get:

- a single deleter object that passes can use to manage iterator invalidation and dead code cleanup

- takes existing complexity that passes were handling manually and move it into this utility 

- begin eliminating the need fro passes to manage InstModCallbacks

I'd love to take this further and begin rewriting passes that call this API, especially now that we have deferred instruction deletion, but I have a lot of other work on core OSSA utilities that is bit-rotting. If someone else can keep going on this that's great.

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -375,7 +387,42 @@ bool tryCheckedCastBrJumpThreading(
 /// cleaning up any dead code resulting from deleting those instructions. Use
 /// this utility instead of
 /// \c recursivelyDeleteTriviallyDeadInstruction.
+///
```
Added comment
```
/// this utility instead of \p recursivelyDeleteTriviallyDeadInstruction
/// as follows:
///   InstructionDeleter deleter;
///   deleter.deleteIfDead();
///   deleter.cleanupDeadInstructions();
```

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -384,20 +431,33 @@ class InstructionDeleter {
+  /// Calls callbacks.notifyWillBeDeleted().
+  bool trackIfDead(SILInstruction *inst);
+
+  /// Force track this instruction as dead. Used to enable the deletion of a
```
"force" always means "ignore side-effects" in this utility.

`  /// Track this instruction as dead even if it has side effects. Used to enable`

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -384,20 +431,33 @@ class InstructionDeleter {
-  void deleteIfDead(SILInstruction *inst);
+  /// If the instruction \p inst is dead, delete it immediately along with its
+  /// destroys and scope-ending uses, and record its operands so that they can
+  /// be cleaned up later.
```
done

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -384,20 +431,33 @@ class InstructionDeleter {
-  /// its operands so that they can be cleaned up later.
-  void deleteIfDead(SILInstruction *inst);
+  /// If the instruction \p inst is dead, delete it immediately along with its
+  /// destroys and scope-ending uses, and record its operands so that they can
```
```
  /// If the instruction \p inst is dead, delete it immediately along with its
  /// destroys and scope-ending uses. If any operand definitions will become
  /// dead after deleting this instruction, track them so they can be deleted
  /// later during cleanUpDeadInstructions().
```

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -411,10 +471,8 @@ class InstructionDeleter {
-  /// callback cannot be used to update instruction iterators since other
-  /// instructions to be deleted remain in the instruction list.
+  /// callbacks.notifyWillBeDeleted will not be called for \p inst but will be
+  /// called for any other instructions that become dead as a result.
```
Rewrote the original comment:
```
  /// Delete the instruction \p inst, ignoring its side effects. If any operand
  /// definitions will become dead after deleting this instruction, track them
  /// so they can be deleted later during cleanUpDeadInstructions(). This
  /// function will add necessary ownership instructions to fix the lifetimes of
  /// the operands of \p inst to compensate for its deletion.
  ///
  /// \pre the function containing \p inst must be using ownership SIL.
  /// \pre the instruction to be deleted must not have any use other than
  /// incidental uses.
  ///
  /// callbacks.notifyWillBeDeleted will not be called for \p inst but will be
  /// called for any other instructions that become dead as a result.
  void forceDeleteAndFixLifetimes(SILInstruction *inst);
```

atrick commented on lib/SIL/IR/SILInstruction.cpp:
```diff
@@ -145,13 +145,15 @@ void SILInstruction::dropAllReferences() {
       succ = nullptr;
     }
   }
-
```
I don't know why an extra line is warranted. It went away during a merge. I put it back.

atrick commented on lib/SIL/IR/SILInstruction.cpp:
```diff
@@ -145,13 +145,15 @@ void SILInstruction::dropAllReferences() {
+  dropNonOperandReferences();
+}
 
+void SILInstruction::dropNonOperandReferences() {
```
This is merely an implementation detail. The instruction deleter was already dropping the operands individually while processing them. So calling `dropAllReferences` after that seemed like essentially a bug to me. Certainly wasteful.

atrick committed:
Update comments for code review.

atrick commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -375,7 +387,42 @@ bool tryCheckedCastBrJumpThreading(
+///
 class InstructionDeleter {
+public:
+  static InstructionDeleter updatingIterator(SILBasicBlock::iterator &iter) {
```
This works exactly like it does today with the onDelete callback. You need to advance the iterator before potentially deleting the current instruction.

I agree it's easy to get wrong. I'm tempted to provide another abstraction, but I think that should be reviewed in a new PR. It's not straightforward. You'll never be able to use a range-based for loop safely with a "non-deleted instruction iterator" because that requires access to the current instruction, which may have been deleted. You could provide a `next`-style iterator, and I do think that's worth doing.

atrick replied:
For now, I'll at least comment `updatingIterator` and add a TODO to design an iterator abstraction.

atrick replied:
@eeckstein Ok, I lied about never using a "range-base" for loop. You can do it as long as you're willing to return nullptr for the current instruction if it was already deleted. I actually think that could be very useful, so went ahead and implemented it:

https://github.com/apple/swift/pull/37669/commits/7b9d8af4bb1df2495f4fc6886ddae5c11f3af420#diff-31e14eaf82b08085838ced1032b31bf8ba537186058636027fb47bbe7cc98f99R2855

Now you can do totally safe iteration:

https://github.com/apple/swift/pull/37669/files#diff-34cd3ed8b9f60f7d19c7762a465e1f1b9eda27644c4bfaae72d65e744d567fbaR480

And it handles new instructions properly, which used to be broken.

atrick commented on lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp:
```diff
@@ -2882,11 +2898,12 @@ static bool eliminateDeadAllocations(SILFunction &fn) {
     while (i != e) {
       // First see if i is an allocation that we can optimize. If not, skip it.
       AllocationInst *alloc = getOptimizableAllocation(&*i);
+      ++i;
```
I had to fix several such cases already. If `getOptimizable` had access to an InstructionDeleter or mutated the SIL, then that would be a concern. I can clarify the code here until we introduce a "non-deleted instruction iterator"

atrick committed:
Add UpdatingInstructionIterator.

Track in-use iterators and update them both when instructions are
deleted and when they are added.

Safe iteration in the presence of arbitrary changes now looks like
this:

    for (SILInstruction *inst : deleter.updatingRange(&bb)) {
      modify(inst);
    }

atrick committed:
Move InstModCallbacks into its own header.

Required to break circular dependence when introducing
UpdatingInstructionIterator.

Also, this is a lot of detail that doesn't belong in the general
InstOptUtils APIs. It's not something people should ever reach for.

eeckstein commented on include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h:
```diff
@@ -0,0 +1,243 @@
+
+namespace detail {
+
+struct UpdatingInstructionIteratorState {
```
Is it really worth separating the state and the iterator into two separate classes and doing reference counting? It adds quite a bit of complexity. Can't you just add references to all iterators into the notification-receiving list?

Also, I'm not sure if this really works. After this code
```
  iterator2 = iterator1
  iterator1++
``` 
both iterators still share the same state, but should refer to different instructions

atrick replied:
Sorry, what I pushed last night was an incomplete mish-mash. I originally had a next-style iterator as an RAII object that registered and deregistered itself. I played with adding an iterator adapter to see if the range-based API would work, but haven't finished. My biggest concern now is that an iterator is inadvertently copied on each iteration. I was considering prohibiting post-increment and keeping the registry efficient. The alternative is dropping the iterator ID and registering a pointer to each iterator. I'm concerned about the overhead, but I'll try that and see how bad it is since there won't be any weird pitfalls with that approach.

atrick replied:
@eeckstein ok, here's the finished UpdatingIterator template. I tested the reverse iterators locally. Everything has been converted over to normal iterator values now. No reference counting, no separate state. We just need to be careful not to create lots of instruction iterator copies on the critical path. So far it isn't a problem.

https://github.com/apple/swift/pull/37669/commits/c05416a17ec602839de97dab88e1b01f890d8221

eeckstein commented on include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h:
```diff
@@ -0,0 +1,243 @@
+struct UpdatingInstructionIteratorState {
+  using Self = UpdatingInstructionIteratorState;
+
+  SILBasicBlock::iterator base;
```
If you make this a template type, it will also work with the reverse_iterator

atrick replied:
Agreed. We should support reverse_iterator.

eeckstein commented on include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h:
```diff
@@ -0,0 +1,243 @@
+
+  void releaseIterator(unsigned id) {
+    if (--iteratorStates[id].refCount == 0)
+      iteratorStates.erase(iteratorStates.begin() + id);
```
That will erase the wrong state if another state was released before with a lower id.

eeckstein commented on include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h:
```diff
@@ -0,0 +1,243 @@
+/// Track iterators that need updating when intructions are added or
+/// deleted. Iterators can be tracked across multiple levels of the call
+/// stack. This registry object must outlive any iterators that it vends.
+class UpdatingInstructionIteratorRegistry {
```
Are you expecting this to be used outside InstructionDeleter?
If not I suggest to merge this with InstructionDeleter (to keep things simple).

atrick replied:
The implementations of InstructionDeleter and UpdatingInstructionIteratorRegistry are unrelated and each complex in their own way. I assure you it's much simpler in a separate class. I do imagine moving it out of InstructionDeleter at some point.

eeckstein replied:
ok

eeckstein commented on include/swift/SILOptimizer/Utils/InstOptUtils.h:
```diff
@@ -373,48 +203,92 @@ bool tryCheckedCastBrJumpThreading(
+/// time. It is the client's responsiblity to invoke notifyWillBeDeletedFunc()
+/// on those explicitly deleted instructions if needed.
+///
+/// TODO: Move this into a separate header.
```
I think it makes sense to do this now instead of adding ` UpdatingInstructionIterator.h` - and put the logic of `UpdatingInstructionIteratorRegistry` directly into `InstructionDeleter` (see my other comment).

atrick replied:
I moved InstModCallbacks and UpdatingInstructionIterator because those classes completely obscured the important InstOptUtils API declarations. Those two classes are implementation details that people never need to look at unless they care about those implementations.

InstructionDeleter is a top-level API that people need to scan all the time. I don't really care if InstructionDeleter goes in a separate header, but that's not necessary and it would still need to be referenced from InstOptUtils.h.

eeckstein replied:
ok

eeckstein commented on lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp:
```diff
@@ -2838,13 +2850,12 @@ static bool optimizeMemoryAccesses(SILFunction &fn) {
   for (auto &bb : fn) {
-    auto i = bb.begin(), e = bb.end();
-    while (i != e) {
+    for (SILInstruction *inst : deleter.updatingRange(&bb)) {
```
nice!

eeckstein commented on include/swift/SILOptimizer/Utils/UpdatingInstructionIterator.h:
```diff
@@ -0,0 +1,243 @@
+///
+/// Or more commonly, directly from the InstructionDeleter:
+///
+///   for (SILInstruction *inst : deleter.iteratorRange(bb)) ...
```
`iteratorRange` -> `updatingRange`

eeckstein reviewed (commented):
Looks much better now!
I still have a few comments about the updating iterator implementation.

atrick commented:
@swift-ci test

atrick commented:
@eeckstein I addressed everything that I wanted to in this PR (this was originally just a bug fix). I'll merge it to unblock the next one, but I'd appreciate if you spot anything else troubling. There's still a lot of work to simplify the passes themselves based on new utilities. Ultimately I think we should have one pass context object to encapsulate stuff that the SILBuilder, InstructionDeleter, UpdatingInstructionIterator, etc. need and have SIL mutation go through that.

eeckstein commented:
Thanks, looks good!
